src/jni_core/src/opendal_schema_history.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::borrow::Cow;
+use std::sync::{Arc, OnceLock};
+
+use bytes::Bytes;
+use jni::objects::{JByteArray, JObject, JString};
+use risingwave_common::config::ObjectStoreConfig;
+use risingwave_common::{DATA_DIRECTORY, STATE_STORE_URL};
+use risingwave_object_store::object::object_metrics::ObjectStoreMetrics;
+use risingwave_object_store::object::{ObjectStoreImpl, build_remote_object_store};
+
+use crate::{EnvParam, JAVA_BINDING_ASYNC_RUNTIME, execute_and_catch, to_guarded_slice};
+
+static OBJECT_STORE_INSTANCE: OnceLock<Arc<ObjectStoreImpl>> = OnceLock::new();
+
+// schema history is internal state, all data is stored under the DATA_DIRECTORY directory.
+fn prepend_data_directory(path: &str) -> String {
+    let data_dir = DATA_DIRECTORY.get().map(|s| s.as_str()).unwrap_or("");
+    if data_dir.is_empty() || path.starts_with(data_dir) {
+        path.to_owned()
+    } else if data_dir.ends_with('/') || path.starts_with('/') {
+        format!("{}{}", data_dir, path)
+    } else {
+        format!("{}/{}", data_dir, path)
+    }
+}
+
+async fn get_object_store() -> Arc<ObjectStoreImpl> {
+    if let Some(store) = OBJECT_STORE_INSTANCE.get() {
+        store.clone()
+    } else {
+        let hummock_url = STATE_STORE_URL.get().unwrap();
+        let object_store = build_remote_object_store(
+            hummock_url.strip_prefix("hummock+").unwrap(),
```

Comment:
This assumes the state store must be hummock, which is not always true, especially in CI and test. Can we fall back to memory schema history if hummock is not used?